### PR TITLE
Upgrade to Calico 2.6.1

### LINKED ISSUE #897: Support Kubernetes network policies
Users need to be able to specify of how groups of pods are allowed to communicate with each other and other network endpoints. This is enabled by three parts:

1. enable the `extensions/v1beta1/networkpolicies` runtime config in your apiserver
2. a controller that acts on Network policies
3. the actual network policies

### LINKED ISSUE #1197: k8s calico network policy has no effect
**Is this a request for help?**: Yes

---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): Issue

---

**What version of acs-engine?**: v0.4.0

---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes

**What happened**: 
Installed a kubernetes cluster using ACS 0.4.0 (kubernetes v1.6.6)
Enabled calico network plugin using,
```json
"kubernetesConfig": {
   "networkPolicy": "calico"
}
```
Tried out the network policy example as mentioned in this link,
https://kubernetes.io/docs/tasks/administer-cluster/declare-network-policy/

kubectl gave an error,
```
error: unable to recognize "test.yaml": no matches for networking.k8s.io/, Kind=NetworkPolicy
```

Changed "networking.k8s.io/v1" to "extensions/v1beta1" and tried again.

This time the network policy has no effect (busybox pod is able to communicate with nginx without issue)

**What you expected to happen**:

Expected the network policy to take effect and stop pod communication from busybox to nginx.

**How to reproduce it** (as minimally and precisely as possible):

Install k8s cluster with ACS v0.4.0 with network plugin calico enabled and try out the example mentioned at this link,
https://kubernetes.io/docs/tasks/administer-cluster/declare-network-policy/

**Anything else we need to know**:

The k8s cluster has calico-node daemonset running.
But it does not have a calico policy controller as mentioned in these links,
https://docs.projectcalico.org/v2.0/getting-started/kubernetes/installation/hosted/hosted
https://docs.projectcalico.org/v2.0/getting-started/kubernetes/installation/hosted/calico.yaml

### LINKED ISSUE #1417: Calico should be updated to v2.5
**Is this an ISSUE or FEATURE REQUEST?** (choose one):

Feature request

**What version of acs-engine?**:

latest

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**

Kubernetes

**Request**:

Calico should be updated to v2.5
​
With Kubernetes deprecation of ThirdPartyResources (TPR) in favor of  Custom Resource Definitions (CRD) Calico has to switch over to CRDs. The v2.5 release makes that switch and requires a migration step for deployments using the Kubernetes Datastore. A migration guide can be found https://github.com/projectcalico/calico/tree/master/upgrade/v2.5

Another thing to be aware of is that CRDs are first available in Kubernetes v1.7. This means that if Calico v2.5 is using the Kubernetes Datastore it must be used with Kubernetes 1.7+, it cannot be used 
with Kubernets 1.6.x or lower.
​
Feel free to reach out to me (@tmjd or erik.stidham on slack) or @ozdanborne on github, Kubernetes slack, or Calico slack to collaborate.

### PR DESCRIPTION
**Pull request by dtzar**
**What this PR does / why we need it**:
Updates calico from  v2.4 Kubernetes Datastore hosted installation to v2.6.1, supporting 1.7+ K8s clusters.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: 
fixes #1417, fixes #1197, fixes #897

```release-note
Updates calico from  v2.4 Kubernetes Datastore hosted installation to v2.6.1, supporting 1.7+ K8s clusters.
```

### CODE REVIEW

**Comment by dtzar** [comment]
The 1.6 failure is expected since calico 2.6 is designed to only work with K8s 1.7 or later.

**Comment by dtzar** [comment]
I found a bug with calico 2.6.0 running on 1.8 ACS engine from Lachie's branch and they fixed it in 2.6.1 
 https://github.com/projectcalico/calico/issues/1131#issuecomment-333281999

**Comment by jackfrancis** [comment]
@dtzar The best way to manage calico going forward given that different versions of calico work better on different versions of Kubernetes is to dynamically assign the image variable string depending on the version of Kubernetes.

The implementation is clunky, but a representative example is the dashboard image. To get started, look at the Kubernetes version-specific constants defined in `pkg/acsengine/const.go`, e.g.:

https://github.com/Azure/acs-engine/blob/master/pkg/acsengine/const.go#L95

The `"dashboard"` key for each version is consumed thusly in `pkg/acsengine/engine.go`:

https://github.com/Azure/acs-engine/blob/master/pkg/acsengine/engine.go#L541

Continuing this example, search for all instances of `kubernetesDashboardSpec` in the codebase and you will be able to follow how the assignments are dispatched along the pasta sauce. Let me know if you have any more questions!

**Comment by lachie83** [comment]
cc @ozdanborne

**Comment by dtzar** [comment]
There are significant changes in 2.6.x which only work with 1.7+ which means that we really need an entirely separate YAML file for the addons.  This is why I [made this change](https://github.com/Azure/acs-engine/pull/1511/commits/538dfee00feed15e19e7350137e96f6f9ba247cd).  

It would be great to know the long-term plan from @ozdanborne on which versions of calico node/cni they plan to keep compatible with what version(s) of K8s.  In looking through the release notes, it seemed like there weren't any significant changes from 2.4.1 --> 2.5.1 so I left the version of calico at 2.4.1 for K8s 1.5 and 1.6.  Please let me know if you believe it should be otherwise.  Once we know the plan/strategy here, then this will help inform the need for when we may (or may not) need to implement the similar strategy to what is done with the dashboard spec as @jackfrancis has mentioned above.

**Comment by dtzar** [comment]
@jchauncey Is there a problem with at least the k8s-1.7-release-e2e test itself?  I did a local acs-engine build and deployed a calico and a non calico 1.7 K8s cluster successfully.

**Comment by jchauncey** [comment]
Yeah I am working on a fix for those tests right now.

File: examples/networkpolicy/README.md [hunk 1]
```diff
@@ -2,4 +2,25 @@
...
+```
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "kubernetesConfig": {
+        "networkPolicy": "calico"
+      }
+```
+
+This deploys the [v2.6.0 release](https://docs.projectcalico.org/v2.6/releases/) of [Kubernetes Datastore Install](https://docs.projectcalico.org/v2.6/getting-started/kubernetes/installation/hosted/kubernetes-datastore/) version of calico with the "Calico policy-only with user-supplied networking" which supports kubernetes ingress policies and has some limitations as denoted on the referenced page.
```
Comment (ozdanborne) [inline examples/networkpolicy/README.md]: nit: `v2.6.1`

Comment (ozdanborne) [reply examples/networkpolicy/README.md]: Also, I don't think this sentence is 100% accurate. The v2.6 release of Calico is applied if using Kubernetes v1.7+. But  [a compatible calico manifest](https://github.com/Azure/acs-engine/blob/master/parts/kubernetesmasteraddons-calico-daemonset1.5.yaml) is applied automatically when using <v1.7.

**Comment by ozdanborne** [comment]
I think we can get this in without worrying that https://github.com/Azure/acs-engine/blob/master/parts/kubernetesmasteraddons-calico-daemonset1.5.yaml is still at Calico v2.4, especially since Calico v2.5 was mainly KDD updates for compatibility with k8s-v1.7.

**Comment by zimmertr** [comment]
```
{
  "apiVersion": "vlabs",
  "properties": {
    "orchestratorProfile": {
      "orchestratorType": "Kubernetes",
      "orchestratorRelease": "1.8",
      "kubernetesConfig": {
        "clusterSubnet": "172.16.38.0/23",
        "networkPolicy": "calico",
        "enableRbac": true
      }
    },
    "masterProfile": {
      "count": 3,
      "dnsPrefix": "dev-eu2-m",
      "vmSize": "Standard_D2_v2",
      "vnetSubnetId": "/subscriptions/REDACTED/resourceGroups/dev-eu2/providers/Microsoft.Network/virtualNetworks/dev-eu2-vnet/subnets/dev-eu2-k8s-subnet",
      "firstConsecutiveStaticIP": "172.16.38.30", 
      "vnetCidr": "172.16.32.0/19"
    },
    "agentPoolProfiles": [
      {
        "name": "agent",
        "count": 3,
        "vmSize": "Standard_D3_v2",
        "vnetSubnetId": "/subscriptions/REDACTED/resourceGroups/dev-eu2/providers/Microsoft.Network/virtualNetworks/dev-eu2-vnet/subnets/dev-eu2-k8s-subnet",
        "availabilityProfile": "AvailabilitySet"
      }
    ],
    "linuxProfile": {
      "adminUsername": "user",
      "ssh": {
        "publicKeys": [
          {
            "keyData": "REDACTED"
          }
        ]
      }
    },
    "servicePrincipalProfile": {
      "clientId": "REDACTED",
      "secret": "REDACTED"
    }
  }
}
```

**Comment by dtzar** [comment]
@zimmertr The newest calico release 2.6.2 has NO changes relevant for Kubernetes, per the release notes.  It is likely some other commit which touched etcd (like PR 1564) caused problems.  Please file a new issue.

**Comment by zimmertr** [comment]
@dtzar thank you for your feedback. I've made a new issue which can be found here: https://github.com/Azure/acs-engine/issues/1621

**Comment by zimmertr** [comment]
@dtzar thank you for your feedback. I've made a new issue which can be found here: #1621 

Can you elaborate on which software you intend for me to revert back to 0.8.0? `etcd`?

**Comment by dtzar** [comment]
Build or download/use the acs-engine binary from this commit/release https://github.com/Azure/acs-engine/releases/tag/v0.8.0

**Pull request merged**